src/vsr/client.zig

```diff
@@ -424,28 +632,63 @@ pub fn Client(comptime StateMachine_: type, comptime MessageBus: type) type {
 
             self.request_timeout.stop();
 
-            if (inflight_operation == .register) {
+            if (inflight_vsr_operation == .register) {
                 assert(self.session == 0);
                 assert(reply.header.commit > 0);
                 self.session = reply.header.commit; // The commit number becomes the session number.
             }
 
-            // We must process the next request before releasing control back to the callback.
-            // Otherwise, requests may run through send_request_for_the_first_time() more than once.
-            if (self.request_queue.head_ptr()) |next_request| {
-                self.send_request_for_the_first_time(next_request.message);
+            // We must start the next request before releasing control back to the callback(s).
+            // Otherwise, requests may never run through send_request_for_the_first_time().
+            self.request_inflight = self.request_queue.head_ptr();
+            if (self.request_inflight != null) {
+                self.send_request_for_the_first_time();
             }
 
-            if (inflight.callback) |callback| {
-                assert(!inflight_operation.vsr_reserved());
+            // Ignore callback processing if the Request was from register().
+            if (inflight.demux.callback == null) {
+                assert(inflight_vsr_operation == .register);
+                assert(inflight.demux_queue.empty());
+                return;
+            }
 
-                callback(
-                    inflight.user_data,
-                    inflight_operation.cast(StateMachine),
-                    reply.body(),
-                );
-            } else {
-                assert(inflight_operation == .register);
+            assert(!inflight_vsr_operation.vsr_reserved());
+            switch (inflight_vsr_operation.cast(StateMachine)) {
+                inline else => |operation| {
+                    var demuxer = StateMachine.DemuxerType(operation).init(
+                        std.mem.bytesAsSlice(StateMachine.Result(operation), reply.body()),
+                    );
+
+                    if (!StateMachine.batch_logical_allowed.get(operation)) {
+                        assert(inflight.demux_queue.empty());
+                    }
+
+                    // Push the inflight's demux to the front of the demux_queue to simplify handling.
+                    inflight.demux.next = inflight.demux_queue.out;
+                    inflight.demux_queue.out = &inflight.demux;
+                    inflight.demux_queue.count += 1;
```

Comment:
It would be self-referential and `request_queue.pop()` returns a (RLS) copy from its array, so `demux_queue_head` in the FIFO would need a fixup after that call.

Reply:
Sadly true 😅
Well, in this case we have the option of acquiring from the DemuxPool in either cases, what do you think? No queue head required!

I think we can't exceed the DemixPool’s capacity the way it is calculated.

Reply:
It could introduce a failure/unreachable in `raw_request` if `get_message()` doesn't pair well with demux nodes, although  that's highly unlikely. `raw_request` is also only used by testing code (aof/cluster) which should already ensure message limits. Ill go ahead.